timeline_shifts_visualization_experiment/01_umap_parameter_sweep.py

```diff
@@ -0,0 +1,275 @@
+# ---
+
+# # Parameter Sweep for UMAP
+
+# The last piece of this puzzle is to generate a visualization to show how a word is shifting through time.
+# The initial analysis here is to see how the original umap clusters words together and figure out the optimal parameters.
+# The next notebook will explore parametric umap, which is a hybrid of umap and neural networks to project data into a low dimensional space.
+
+# +
+import itertools
+from pathlib import Path
+import re
+
+from gensim.models import Word2Vec, KeyedVectors
+import numpy as np
+import pandas as pd
+import plotnine as p9
+import plydata as ply
+import plydata.tidy as ply_tdy
+import tqdm
+import umap
+# -
+
+# # Unaligned Models
+
+# ## Load the paths for each year
+
+# The goal here is to grab the odd years unaligned and see if umap can tease apart odd years.
+
+unaligned_word_models = list(
+    Path("../multi_model_experiment/output/models").rglob("*model")
+)
+unaligned_word_models = sorted(unaligned_word_models, key=lambda x: x.stem)
+unaligned_word_model_filter = list(
+    filter(lambda x: "_0" in x.stem, unaligned_word_models)
+)
+
+word_freq_count_cutoff = 5
+odd_year_subset = unaligned_word_model_filter[11:-1:2]
+
+training_unaligned_word_model_map = dict()
+for word_file in tqdm.tqdm(odd_year_subset):
+    model = Word2Vec.load(str(word_file)).wv
+    training_unaligned_word_model_map[word_file.stem.split("_")[0]] = dict(
+        model=model,
+        cutoff_index=min(
+            map(
+                lambda x: 999999
+                if model.get_vecattr(x[1], "count") > word_freq_count_cutoff
+                else x[0],
+                enumerate(model.index_to_key),
+            )
+        ),
+    )
+
+words_to_visualize = []
+token_character_list = []
+year_labels_list = []
+
+for year in tqdm.tqdm(training_unaligned_word_model_map):
+    model = training_unaligned_word_model_map[year]["model"]
+    word_subset_matrix = model[
+        model.index_to_key[: training_unaligned_word_model_map[year]["cutoff_index"]]
+    ]
+    print((year, word_subset_matrix.shape))
+    words_to_visualize.append(word_subset_matrix)
+    token_character_list += list(
+        map(
+            lambda x: re.escape(x),
+            model.index_to_key[
+                : training_unaligned_word_model_map[year]["cutoff_index"]
+            ],
+        )
+    )
+    year_labels_list += [year] * len(
+        model.index_to_key[: training_unaligned_word_model_map[year]["cutoff_index"]]
+    )
+
+training_unaligned_words = np.vstack(words_to_visualize)
+training_unaligned_words
+
+# ## Run the Parameter Sweep
+
+n_neighbors = [10, 25, 50, 75]
+epoch_multiplier = [1]
+param_output = []
+
+for neighbors, epoch_m in itertools.product(n_neighbors, epoch_multiplier):
+    file_name = f"output/umap_parameter_sweep/unaligned/neighbors({neighbors})_epochs({epoch_m})"
+
+    if Path(f"{file_name}.tsv").exists():
+        continue
+
+    print(neighbors, epoch_m)
+
+    model = umap.UMAP(
```

Comment:
what is the dimension of your data? It might be useful to do a PCA reduction first if it's very high dimensional in order to ensure that UMAP reliable captures the bulk of the important information.

Reply:
>what is the dimension of your data? It might be useful to do a PCA reduction first if it's very high dimensional in order to ensure that UMAP reliable captures the bulk of the important information.

300 dimensions. I haven't considered that approach. May look into how that would affect these results. looking at the datapoints UMAP appears to work as there are hidden clusters in the plots. Thanks for the tip!

Reply:
it may not be needed, just a common complaint in single-cell UMAP usage.